### ISSUE TITLE
[Suggestion] Create a rewrite branch

### ISSUE THREAD

**Issue by Faunsce** [Drewol/unnamed-sdvx-clone#229]
As mentioned in #153 , A ground up rewrite is a very good idea and with this we can use old code as reference to heavily improve the design. This would allow people to make pull requests with the explicit purpose of building upon the rewrite.

**Comment by Drewol** [comment]
I can create a branch that people can PR rewrites to but I will likely not contribute much to the branch myself. I personally don't think a full ground up rewrite is necessary but I do think that some of the modules could be improved a fair bit by rewriting them.

**Comment by Faunsce** [comment]
I fully understand that you do not have the time but providing the option for the community to help rewrite the code in a clear place is in my opinion a good idea. You may be right and a ground up rewrite may not be necesarry however, after reading through some of the issues posted here, a large amount of code seems to be in need of rewrites. I guess its up to the community including yourself to decide which is best.

**Comment by BonusPlay** [comment]
There is a rewrite branch on my repo [here](https://github.com/BonusPlay/unnamed-sdvx-clone).

**Comment by Faunsce** [comment]
@BonusPlay Jun 4 was your last commit so i was under the impression the repo was dead, also you are refactoring not rewriting which is entirely different and could yield for a more much easily developed code base.

**Comment by BonusPlay** [comment]
I mean, if you've got enough free time on your hands then feel free to go ahead and write it from scratch. Since it's a rewrite it souls be in a different repo.

IMO this project requires a rewrite, but it's not as small as you may think (it would take more than a month of full-time work to do). Since none has that much time to spend on an open source hobby project, I believe a better approach is to refactor with proper interfaces (like `ITexture` which allows to easily swap implementations of whole parts of code (so people wanting to contribute need to deal with smaller pieces at once).

The reason I haven't been working on refactoring is that:
 - lately I got myself a job
 - it's a bit underwhelming when you spend ~50h on refactoring just to see another 250h ahead of you and no progress (as in no new functionality) to justify more work
 - project felt a bit too big to refactor alone (actually I got 2 PRs from @nullableVoidPtr which gave me a bit motivation boost to continue working on)

**Comment by Faunsce** [comment]
I see, and first off thank you for explaining. I'm a relatively new C++ developer and while I spend like 50% of my waking hours coding/learning C++, 5 months of experience isn't a lot so laying it out like this makes it easier for someone such as myself to understand. My two biggest goals here are to 1. respect the dev's who've worked on this and 2. not make an ass of myself. I'll definetly read into the code you have so far and see what i can do. I've never really refactored, only rebuilt so I'll see what I can do. I am also a university student so I feel you.

**Comment by Hinara** [comment]
Things todo while rewriting (that came to my mind) which could be great:
- Choose a dependency model (submodule or package manager)
- Add unit tests ! (Really important)
- Maybe define the coding style to use to be homogeneous
- Clean code and make it more homogeneous too

**Comment by Drewol** [comment]
I think this should probably be on hold till after 1.0, as like maybe a target for 2.0, as there will be not insignificant additions and changes to the current code till 1.0.

But this issue/thread can be the place for making plans for that.

**Comment by Faunsce** [comment]
I definetly support the idea of using the time between now and gold release to plan out the major api changes for 2.0. If we go with a package manager, I've had a really great time with Conan, and also removing things like the hard packages lua library can really help with things like the download size and repository pollution.

**Comment by Hinara** [comment]
We should compare Conan and vcpkg for our use case I think.
Some instruction from the readme already use vcpkg, and I know that under conan you may need to add remotes to be able to have all necessary package.

**Comment by Hinara** [comment]
Maybe some branch should also be deleted

Done (already merged) ?
build-system-changes (c32c551e62e387f71d2b7ab5be549ef406c752b9)
feat/hiddenmod (636ddc6d111e4276e22ec1e6ab6bf4a68758a99a)
feat/lobbyowner (6a965c42027e791af5a29a8469e410f24aba3590)
feat/multiportrait (debc4914896559359377a7fa343c4be7071e9542)
feat/scoresync (5e42d896522ecfe7ad3cfe6435c8b6533f0b48ee)
miniaudio (b319f60cb832f592e0c7a1af3011ee76dde7e6bc)
nanovg (a62be232f6dfbefe4384439d45968e8a27d13f0f)
skin-http (a0e6fe85290720a9713a24627ce188a567eabf70)
skin-settings (126bc710ac9f6495884787371f0a670393b4ffba)
skin_animation (d0b004654225450f09fd59ee2e56c02ba7d3fcbc)

Dead ?
embedded (last update 9529595afbfd288e7e5c150b8668b746e33839f3 7/28/2018)
experimental (last update 4f161e6fa82c96c1ff5e52a84cd1cdcb98eaa06a 10/16/2017)

**Comment by Drewol** [comment]
Yeah the merged ones I had just forgotten to delete, experimental was completely dead so i deleted that but I'll keep embedded around till I've managed to get the master to build and run on a raspberry pi (OpenGL ES)

**Comment by Faunsce** [comment]
@Hinara for conan, you basically only have to add one remote called bincrafters as that group has packaged almost every major library out there. I just checked the remotes and i believe that every package we'd need exist under either bincrafters and or the official conan remote.
